lightning-data-persister/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+use lightning::ln::data_persister::ChannelDataPersister;
+use lightning::chain::transaction::OutPoint;
+use lightning::chain::keysinterface::ChannelKeys;
+use lightning::chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr};
+use lightning::util::ser::{Writeable, Readable};
+use bitcoin::hash_types::{BlockHash, Txid};
+use bitcoin::hashes::hex::{ToHex, FromHex};
+use std::collections::HashMap;
+use std::fs;
+use std::io::{Error, ErrorKind, Cursor};
+use std::marker::PhantomData;
+
+/// LinuxPersister can persist channel data on disk on Linux machines, where
+/// each channel's data is stored in a file named after its outpoint.
+pub struct LinuxPersister<ChanSigner: ChannelKeys + Readable + Writeable> {
+	path_to_channel_data: String,
+	phantom: PhantomData<ChanSigner>, // TODO: is there a way around this?
+}
+
+impl<ChanSigner: ChannelKeys + Readable + Writeable> LinuxPersister<ChanSigner> {
+	/// Initialize a new LinuxPersister and set the path to the individual channels'
+	/// files.
+	pub fn new(path_to_channel_data: String) -> Self {
+		return Self {
+			path_to_channel_data,
+			phantom: PhantomData,
+		}
+	}
+
+	fn get_full_filepath(&self, funding_txo: OutPoint) -> String {
+		format!("{}/{}_{}", self.path_to_channel_data, funding_txo.txid.to_hex(), funding_txo.index)
+	}
+
+	fn write_channel_data(&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChanSigner>) -> std::io::Result<()> {
+		// Do a crazy dance with lots of fsync()s to be overly cautious here...
+		// We never want to end up in a state where we've lost the old data, or end up using the
+		// old data on power loss after we've returned
+		// Note that this actually *isn't* enough (at least on Linux)! We need to fsync an fd with
+		// the containing dir, but Rust doesn't let us do that directly, sadly. TODO: Fix this with
```

Comment:
Definitely should find a way to do the best we can w.r.t. the Linux calls to persist data. 

However, that is still only going to guarantee that data is passed to the drive and it is up to the drive manufacturers to do the right thing which they often times don't do. Especially in consumer-grade hardware.

Any user actually doing persistence is responsible for validating their entire storage/os stack to ensure the writes are persistent. If we expect most users to use the default plugin that LDK provides, it should come with a lot of warnings to make sure people know what they are getting.

Reply:
> However, that is still only going to guarantee that data is passed to the drive and it is up to the drive manufacturers to do the right thing which they often times don't do. Especially in consumer-grade hardware.

> Any user actually doing persistence is responsible for validating their entire storage/os stack to ensure the writes are persistent. 

I'll add a comment about that. 

Lmk if you have any more specifics on how the LinuxPersister could be made more robust. Hm, I also can't find any docs on why the `sync_all` isn't enough and `fsync` on the parent dir is required.

Reply:
So the basic idea is that sync_all() syncs the file data and metadata, but not the directory so on POSIX systems a fsync(parent_dir) is required to ensure both data structures are updated. I found a quick article that highlights it:

http://blog.httrack.com/blog/2013/11/15/everything-you-always-wanted-to-know-about-fsync/

And a more in-depth one:
https://lwn.net/Articles/457667/

I did a bit of digging on what other people are doing and it looks like external crates. I found a few old discussions in the rust community about it, but couldn't find a bug to point to that we can add here to help future readers.

The closest I could come is: `https://github.com/rust-lang/rust/issues/32255` where they talked about the issue and the need for a higher-level abstraction for atomic writes that needs an RFC. Helpful, right?

Reply:
Lol, very helpful. I added fsyncing on the parent dir in ` Make writes atomic on *nix platforms.`